runtime/autoload/provider/pythonx.vim

```diff
@@ -22,96 +24,122 @@ function! provider#pythonx#Require(host) abort
-  return ['', 'provider/pythonx: Could not load Python '.a:major_version.":\n".join(errors, "\n")]
+  return ['', '', errors]
 endfunction
 
 " Returns array: [prog_exitcode, prog_version]
 function! s:import_module(prog, module) abort
   let prog_version = system([a:prog, '-c' , printf(
         \ 'import sys; ' .
         \ 'sys.path = list(filter(lambda x: x != "", sys.path)); ' .
-        \ 'sys.stdout.write(str(sys.version_info[0]) + "." + str(sys.version_info[1])); ' .
+        \ 'sys.stdout.write(".".join(str(x) for x in sys.version_info[:3])); ' .
         \ 'import pkgutil; ' .
-        \ 'exit(2*int(pkgutil.get_loader("%s") is None))',
+        \ 'sys.exit(2*int(pkgutil.get_loader("%s") is None))',
         \ a:module)])
   return [v:shell_error, prog_version]
 endfunction
 
-" Returns array: [was_success, error_message]
-function! provider#pythonx#CheckForModule(prog, module, major_version) abort
-  let prog_path = exepath(a:prog)
-  if prog_path ==# ''
-    return [0, a:prog . ' not found in search path or not executable.']
-  endif
+function! s:satisfies_version(target, ref_major, ref_minimum) abort
+  return a:target =~# '^' . a:ref_major && luaeval('vim.cmp_versions(_A[1], _A[2]) >= 0', [a:target, a:ref_minimum])
+endfunction
 
-  let min_version = (a:major_version == 2) ? '2.6' : '3.3'
+" Returns array: [was_success, python_version, error_message]
+function! provider#pythonx#CheckForModule(prog_path, module, major_version) abort
+  let min_version = get(s:, 'min_python'.a:major_version.'_version')
 
   " Try to load module, and output Python version.
   " Exit codes:
   "   0  module can be loaded.
   "   2  module cannot be loaded.
   "   Otherwise something else went wrong (e.g. 1 or 127).
-  let [prog_exitcode, prog_version] = s:import_module(a:prog, a:module)
-
-  if prog_exitcode == 2 || prog_exitcode == 0
-    " Check version only for expected return codes.
-    if prog_version !~ '^' . a:major_version
-      return [0, prog_path . ' is Python ' . prog_version . ' and cannot provide Python '
-            \ . a:major_version . '.']
-    elseif prog_version =~ '^' . a:major_version && prog_version < min_version
-      return [0, prog_path . ' is Python ' . prog_version . ' and cannot provide Python >= '
-            \ . min_version . '.']
-    endif
+  let [prog_exitcode, prog_version] = s:import_module(a:prog_path, a:module)
+
+  " Check version only for expected return codes.
+  if (prog_exitcode == 2 || prog_exitcode == 0) && !s:satisfies_version(prog_version, a:major_version, min_version)
+    return [0, '', a:prog_path . ' is Python ' . prog_version . ' and cannot provide Python '
+          \ . a:major_version . ' >= ' . min_version . '.']
   endif
 
   if prog_exitcode == 2
-    return [0, prog_path.' does not have the "' . a:module . '" module. :help provider-python']
-  elseif prog_exitcode == 127
-    " This can happen with pyenv's shims.
-    return [0, prog_path . ' does not exist: ' . prog_version]
```

Comment:
I wouldn't special case problems with `pyenv` shims, the error output should be enough for the user to determine what's going on, especially since Neovim's addition to it is just a cryptic "/foo/bar/python does not exist".